src/cache/src/index.ts

```diff
@@ -347,36 +355,99 @@ export class Cache extends LRUCache<
-    const tmp = dir + '/.' + base + '.' + this.random
-    const keyTmp = dir + '/.' + keyFile + '.' + this.random
-    const writeData =
-      intFile ?
-        // try to just link into the temp location, rather than write
-        // may fail with ENOENT or EEXIST, in which case, write normally
-        link(intFile, tmp).catch(() => writeFile(tmp, val))
-        // don't have it by integrity, write the new entry
-      : writeFile(tmp, val)
-    return Promise.all([
-      writeData.then(() => rename(tmp, path)),
-      writeFile(keyTmp, key).then(() =>
-        rename(keyTmp, path + '.key'),
-      ),
-    ]).then(() => {
-      if (intFile) {
-        return link(path, intFile)
+    const base = basename(valFile)
+    const keyFile = `${base}.key`
+    const tmp = join(dir, `.${base}.${this.random}`)
+    const keyTmp = join(dir, `.${keyFile}.${this.random}`)
+
+    // Always write the key file in parallel with other operations
+    const writeKeyP = writeFile(keyTmp, key).then(() =>
+      rename(keyTmp, valFile + '.key'),
+    )
+
+    // We might not write the val file at all so we dont want to
+    // kick off this promise until we know we need to.
+    const writeVal = () =>
+      writeFile(tmp, val).then(() => rename(tmp, valFile))
+
+    if (!intFile) {
+      // No integrity provided, just write the value and key files
+      await Promise.all([writeKeyP, writeVal()])
+      return
+    }
+
+    // Handle different integrity scenarios
+
+    const intStats = await success(stat(intFile))
+    if (!intStats) {
+      // Integrity file doesn't exist yet
+      // Write the value file and then link to a new integrity file
+      await Promise.all([
+        writeKeyP,
+        writeVal().then(() => link(valFile, intFile)),
+      ])
+      return
+    }
+
+    // Check if the integrity file and value file are the same size
+    // This is a fast shortcut to check if the content is likely the same
+    const sameSize = intStats.size === val.length
+
+    const valStats = await success(stat(valFile))
+    if (!valStats && sameSize) {
+      // Path doesn't exist but integrity is already what we want
+      // Try to link integrity file to val file
+      // If linking fails, continue on
+      if ((await success(link(intFile, valFile))) !== null) {
+        await writeKeyP
+        return
       }
-    })
+    }
+
+    const linked =
+      intStats.ino === valStats?.ino && intStats.dev === valStats.dev
+
+    if (linked && sameSize) {
+      // Integrity and val file are already linked and the same
+      // So nothing to do except write the key file
+      await writeKeyP
+      return
+    }
+
+    if (!linked && sameSize) {
+      // Integrity and val file are not linked but same size
+      // so its likely the same content so force link them
+      // because thats faster than writing the value file.
+      if (
+        (await success(
+          unlink(valFile).then(() => link(intFile, valFile)),
+        )) !== null
```

Comment:
@isaacs Calling out this block specifically because it should be the only difference from when we went over this in pairing.

Now if the value and integrity files are not linked, but are likely the same, it tries to force them to be linked. 

This should (almost?) always be what we want. The alternative of only calling `link(intFile, valFile)` without first unlinking will always reject with `EEXIST` which renders this block meaningless.

Reply:
Correct, in this case you do know that you need to unlink the valFile, because you know it exists.

Reply:
Oh, actually, this should be an atomic link+rename, though, not an unlink and link, so that it's one fs op and doesn't ever have the file go missing.

Need something like this:

```ts
async #linkAtomic (src: string, dest: string) {
  const tmp = `${dest}.${this.#random}${this.#i++}`
  await link(src, tmp).then(() => rename(tmp, dest))
}
```

```suggestion
          this.#linkAtomic(intFile, valFile)),
        )) !== null
```